signac_dashboard/dashboard.py

```diff
@@ -86,6 +87,15 @@ def __init__(self, config={}, project=None, modules=[]):
         self.config = config
         self.modules = modules
 
+        keyfunc = lambda m: m.context
+        grouped = groupby(sorted(modules, key=keyfunc), key=keyfunc)
+        modules_by_context = dict()
+        for context_key, context_group in grouped:
+            modules_by_context[context_key] = [
+                m for i, m in enumerate(context_group) if m.enabled
+            ]
+        self.modules_by_context = modules_by_context
```

Comment:
Use `{}` instead of `dict()` (literals are faster and better style; I thought pyupgrade enforced this but maybe it's situational).

No need to `enumerate` if the index `i` is unused.

This should be at the end of the `_prepare` function, since that can remove modules if their registration fails. Alternatively, we should remove `_prepare` and move all its contents to the `__init__` function. Either way, this needs to be executed after module registration in case registration fails and modules are removed, and it needs to start from `self.modules` rather than the user-provided `modules`.

Can the `self.modules_by_context` be private? `self._modules_by_context`

```suggestion
        keyfunc = lambda m: m.context
        grouped = groupby(sorted(self.modules, key=keyfunc), key=keyfunc)
        modules_by_context = {}
        for context_key, context_group in grouped:
            modules_by_context[context_key] = [
                m for m in context_group if m.enabled
            ]
        self.modules_by_context = modules_by_context
```

Reply:
That API proposal is interesting to consider but I’d like to avoid that change in this PR. One of my hesitations is that the use of a dict permits arbitrary keys and doesn’t constrain inputs to the allowable context names like a class constructor or free function might do. However, that’s also somewhat true for the module context argument itself. Let’s revisit later.